Source/NSFontCollection.m

```diff
@@ -0,0 +1,533 @@
+	    }
+
+	  de = [fm enumeratorAtPath: dir];
+	  while ((file = [de nextObject])) 
+	    {
+	      if ([[file pathExtension] isEqualToString: @"collection"])
+		{
+		  NSString	*name;
+		  name = [file stringByDeletingPathExtension];
+		  newCollection = [NSKeyedUnarchiver unarchiveObjectWithFile: [dir stringByAppendingPathComponent: file]];
+                  if (newCollection != nil)
+                    {
+                      [_availableFontCollections addObject: newCollection];
+                      RELEASE(newCollection);
+                    }
+                }
+	    }
+	}  
+      /*
+      if (defaultSystemFontCollection != nil)
+        {
+	  [_availableFontCollections addObject: defaultSystemFontCollection];
+	}
+      */
+      [_fontCollectionLock unlock];
+    }
+}
+
+/*
+ * Writing and Removing Files
+ */
+- (BOOL) _writeToFile: (NSString *)path
+{
+  NSFileManager *fm = [NSFileManager defaultManager];
+  NSString      *tmpPath;
+  BOOL          isDir;
+  BOOL          success;
+  BOOL          path_is_standard = YES;
+
+  /*
+   * We need to initialize before saving, to avoid the new file being 
+   * counted as a different collection thus making it appear twice
+   */
+  [NSFontCollection _loadAvailableFontCollections];
+
+  if (path == nil)
+    {
+      NSArray	*paths;
+
+      // FIXME the standard path for saving font collections
+      paths = NSSearchPathForDirectoriesInDomains(NSLibraryDirectory,
+	NSUserDomainMask, YES);
+      if ([paths count] == 0)
+	{
+	  NSLog (@"Failed to find Library directory for user");
+	  return NO;	// No directory to save to.
+	}
+      path = [[paths objectAtIndex: 0]
+	stringByAppendingPathComponent: @"FontCollections"]; 
+      isDir = YES;
+    }
+  else
+    {
+      [fm fileExistsAtPath: path isDirectory: &isDir];
+    }
+
+  if (isDir)
+    {
+      ASSIGN (_fullFileName, [[path stringByAppendingPathComponent: _name] 
+        stringByAppendingPathExtension: @"collection"]);
+    }
+  else // it is a file
+    {
+      if ([[path pathExtension] isEqual: @"collection"] == YES)
+	{
+	  ASSIGN (_fullFileName, path);
+	}
+      else
+	{
+	  ASSIGN (_fullFileName, [[path stringByDeletingPathExtension]
+	    stringByAppendingPathExtension: @"collection"]);
+	}
+      path = [path stringByDeletingLastPathComponent];
+    }
+
+  // Check if the path is a standard path
+  if ([[path lastPathComponent] isEqualToString: @"FontCollections"] == NO)
+    {
+      path_is_standard = NO;
+    }
+  else 
+    {
+      tmpPath = [path stringByDeletingLastPathComponent];
+      if (![NSSearchPathForDirectoriesInDomains(NSLibraryDirectory,
+	NSAllDomainsMask, YES) containsObject: tmpPath])
+	{
+	  path_is_standard = NO;
+	}
+    }
+
+  /*
+   * If path is standard and it does not exist, try to create it.
+   * System standard paths should always be assumed to exist; 
+   * this will normally then only try to create user paths.
+   */
+  if (path_is_standard && ([fm fileExistsAtPath: path] == NO))
```

Comment:
Shouldn't we also create a non existing non standard path? Otherwise writing the archive will fail.